--- samtools.orig/Makefile	2015-08-12 14:11:56.000000000 +0100
+++ samtools/Makefile	2015-08-12 14:30:25.947902991 +0100
@@ -23,7 +23,6 @@
 
 CC       = gcc
 CPPFLAGS =
-CFLAGS   = -g -Wall -O2
 LDFLAGS  =
 LIBS     =
 
@@ -38,8 +37,14 @@
             faidx.o dict.o stats.o stats_isize.o bam_flags.o bam_split.o \
             bam_tview.o bam_tview_curses.o bam_tview_html.o bam_lpileup.o
 
+EBROOTZLIB=../zlib-1.2.8
+INCLUDES=   -I$(EBROOTNCURSES)/include -I$(EBROOTZLIB)/include -I. -I$(HTSDIR)
 EXTRA_CPPFLAGS = $(DFLAGS) -I. -I$(HTSDIR)
-LIBCURSES=  -lcurses # -lXCurses
+
+CFLAGS=$(INCLUDES) -g -Wall -O2
+
+LIBCURSES=  -L$(EBROOTNCURSES)/lib -lcurses # -lXCurses
+ZLIB=		-L$(EBROOTZLIB) -lz
 
 prefix      = /usr/local
 exec_prefix = $(prefix)
@@ -124,7 +129,7 @@
 	$(AR) -csru $@ $(LOBJS)
 
 samtools: $(AOBJS) libbam.a $(HTSLIB)
-	$(CC) -pthread $(LDFLAGS) -o $@ $(AOBJS) libbam.a $(HTSLIB) $(LIBCURSES) -lm -lz $(LIBS)
+	$(CC) -pthread $(LDFLAGS) -o $@ $(AOBJS) libbam.a $(HTSLIB) $(LIBCURSES) -lm $(ZLIB) $(LIBS)
 
 bam_h = bam.h $(htslib_bgzf_h) $(htslib_sam_h)
 bam2bcf_h = bam2bcf.h $(htslib_vcf_h) errmod.h
@@ -196,31 +201,31 @@
 
 
 test/merge/test_bam_translate: test/merge/test_bam_translate.o test/test.o $(HTSLIB)
-	$(CC) -pthread $(LDFLAGS) -o $@ test/merge/test_bam_translate.o test/test.o $(HTSLIB) -lz $(LIBS)
+	$(CC) -pthread $(LDFLAGS) -o $@ test/merge/test_bam_translate.o test/test.o $(HTSLIB) $(ZLIB) $(LIBS)
 
 test/merge/test_pretty_header: test/merge/test_pretty_header.o $(HTSLIB)
-	$(CC) -pthread $(LDFLAGS) -o $@ test/merge/test_pretty_header.o $(HTSLIB) -lz $(LIBS)
+	$(CC) -pthread $(LDFLAGS) -o $@ test/merge/test_pretty_header.o $(HTSLIB) $(ZLIB) $(LIBS)
 
 test/merge/test_rtrans_build: test/merge/test_rtrans_build.o $(HTSLIB)
-	$(CC) -pthread $(LDFLAGS) -o $@ test/merge/test_rtrans_build.o $(HTSLIB) -lz $(LIBS)
+	$(CC) -pthread $(LDFLAGS) -o $@ test/merge/test_rtrans_build.o $(HTSLIB) $(ZLIB) $(LIBS)
 
 test/merge/test_trans_tbl_init: test/merge/test_trans_tbl_init.o $(HTSLIB)
-	$(CC) -pthread $(LDFLAGS) -o $@ test/merge/test_trans_tbl_init.o $(HTSLIB) -lz $(LIBS)
+	$(CC) -pthread $(LDFLAGS) -o $@ test/merge/test_trans_tbl_init.o $(HTSLIB) $(ZLIB) $(LIBS)
 
 test/split/test_count_rg: test/split/test_count_rg.o test/test.o $(HTSLIB)
-	$(CC) -pthread $(LDFLAGS) -o $@ test/split/test_count_rg.o test/test.o $(HTSLIB) -lz $(LIBS)
+	$(CC) -pthread $(LDFLAGS) -o $@ test/split/test_count_rg.o test/test.o $(HTSLIB) $(ZLIB) $(LIBS)
 
 test/split/test_expand_format_string: test/split/test_expand_format_string.o test/test.o $(HTSLIB)
-	$(CC) -pthread $(LDFLAGS) -o $@ test/split/test_expand_format_string.o test/test.o $(HTSLIB) -lz $(LIBS)
+	$(CC) -pthread $(LDFLAGS) -o $@ test/split/test_expand_format_string.o test/test.o $(HTSLIB) $(ZLIB) $(LIBS)
 
 test/split/test_filter_header_rg: test/split/test_filter_header_rg.o test/test.o $(HTSLIB)
-	$(CC) -pthread $(LDFLAGS) -o $@ test/split/test_filter_header_rg.o test/test.o $(HTSLIB) -lz $(LIBS)
+	$(CC) -pthread $(LDFLAGS) -o $@ test/split/test_filter_header_rg.o test/test.o $(HTSLIB) $(ZLIB) $(LIBS)
 
 test/split/test_parse_args: test/split/test_parse_args.o test/test.o $(HTSLIB)
-	$(CC) -pthread $(LDFLAGS) -o $@ test/split/test_parse_args.o test/test.o $(HTSLIB) -lz $(LIBS)
+	$(CC) -pthread $(LDFLAGS) -o $@ test/split/test_parse_args.o test/test.o $(HTSLIB) $(ZLIB) $(LIBS)
 
 test/vcf-miniview: test/vcf-miniview.o $(HTSLIB)
-	$(CC) -pthread $(LDFLAGS) -o $@ test/vcf-miniview.o $(HTSLIB) -lz $(LIBS)
+	$(CC) -pthread $(LDFLAGS) -o $@ test/vcf-miniview.o $(HTSLIB) $(ZLIB) $(LIBS)
 
 test_test_h = test/test.h $(htslib_sam_h)
 
@@ -239,22 +244,22 @@
 # misc programs
 
 misc/ace2sam: misc/ace2sam.o
-	$(CC) $(LDFLAGS) -o $@ misc/ace2sam.o -lz $(LIBS)
+	$(CC) $(LDFLAGS) -o $@ misc/ace2sam.o $(ZLIB) $(LIBS)
 
 misc/maq2sam-short: misc/maq2sam-short.o
-	$(CC) $(LDFLAGS) -o $@ misc/maq2sam-short.o -lz $(LIBS)
+	$(CC) $(LDFLAGS) -o $@ misc/maq2sam-short.o $(ZLIB) $(LIBS)
 
 misc/maq2sam-long: misc/maq2sam-long.o
-	$(CC) $(LDFLAGS) -o $@ misc/maq2sam-long.o -lz $(LIBS)
+	$(CC) $(LDFLAGS) -o $@ misc/maq2sam-long.o $(ZLIB) $(LIBS)
 
 misc/md5fa: misc/md5fa.o $(HTSLIB)
-	$(CC) $(LDFLAGS) -o $@ misc/md5fa.o $(HTSLIB) -lz $(LIBS)
+	$(CC) $(LDFLAGS) -o $@ misc/md5fa.o $(HTSLIB) $(ZLIB) $(LIBS)
 
 misc/md5sum-lite: misc/md5sum-lite.o $(HTSLIB)
 	$(CC) $(LDFLAGS) -o $@ misc/md5sum-lite.o $(HTSLIB) $(LIBS)
 
 misc/wgsim: misc/wgsim.o
-	$(CC) $(LDFLAGS) -o $@ misc/wgsim.o -lm -lz $(LIBS)
+	$(CC) $(LDFLAGS) -o $@ misc/wgsim.o -lm $(ZLIB) $(LIBS)
 
 misc/ace2sam.o: misc/ace2sam.c $(HTSDIR)/htslib/kstring.h $(HTSDIR)/htslib/kseq.h
 misc/md5fa.o: misc/md5fa.c $(htslib_hts_h) $(HTSDIR)/htslib/kseq.h
